### PR TITLE
[fix] group tx, ctx, store flow

### PR DESCRIPTION
**Pull request by amityadav0**
- Fix group transaction flow
- Fix Clear SSC
- Fix tests

### CODE REVIEW

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:702]: Can we move `this.processTransactions()` to `Ctx` class (if we are operating entirely on `ctx` in this function) ?

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:642]: Please add a comment regarding error no. `1007`.

File: packages/runtime/src/ctx.ts
Comment (ratik21) [inline packages/runtime/src/ctx.ts:42]: Can you please explain why we're using `runtime.ctx` here? We are in `Ctx` class but we are operating on `runtime.ctx.state` here. Shouldn't this be `this.state` (which would be a deepCopy of `runtime.store`)?

File: examples/crowdfunding/test/crowdfundingTest.js
Comment (ratik21) [inline examples/crowdfunding/test/crowdfundingTest.js:111]:
```suggestion
    const fundEscrowParam = {
```

File: examples/crowdfunding/test/crowdfundingTest.js
Comment (ratik21) [inline examples/crowdfunding/test/crowdfundingTest.js:119]:
```suggestion
    runtime.executeTx(fundEscrowParam);
```

File: packages/runtime/src/ctx.ts
Comment (ratik21) [inline packages/runtime/src/ctx.ts:108]:
```suggestion
   * Deduct transaction fee from sender account.
```

File: packages/runtime/src/ctx.ts
Comment (ratik21) [inline packages/runtime/src/ctx.ts:110]:
```suggestion
   * @param index Index of transaction in tx group
```

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:642]: ok, this is because we only closeout if error is rejected by logic, other than that we should throw error,
we were missing this earlier. will add in code also

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [reply packages/runtime/src/ctx.ts:42]: Because it is not required to make a deepcopy of a deepcopy.
think about the flow, first we make a deepcopy and send it to interpreter(interpreter works on runtime.ctx) if the execution fails at some point we don't update the runtime.store.

same is the case with processTransaction, we process transactions on deepcopy of store, if it fails we don;t update store with ctx, else we update store.

Comment (amityadav0) [reply packages/runtime/src/ctx.ts:42]: let me know it that makes sense.

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:702]: yes

File: packages/runtime/src/ctx.ts
Comment (ratik21) [reply packages/runtime/src/ctx.ts:42]: No i understand about the deepCopy, i am talking about the intuitiveness here (using  `runtime.ctx.state`) inside a `Ctx` class.
When i was talking about using `Ctx` as a class, i was referring to `runtime.ctx` only (not `ctxOp`). Basically having `runtime.ctx = new Ctx(deepCopy(this.store), tx, gtxs, args..)`. Class `Ctx` would be of type `Context` in `/runtime/types` only. Interpreter would have used this class functions (instead of having local interpreter functions). This way the separation b/w `ctx & store` would be better also.

For here, maybe during initialization we can do `this.state = runtime.ctx.state` in the constructor? Otherwise it seems a little confusing to me. We can keep as it is if you think otherwise.

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts:642]: ok

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [reply packages/runtime/src/ctx.ts:42]: ok, do you mean ``` export interface Context {
  state: State
  tx: Txn // current txn
  gtxs: Txn[] // all transactions
  args: Uint8Array[]
  transferAlgo()
  transferAsset()
}```

we will have these in types and we will have this class as Context. right ?
yes, this seems better.

ctx = deepCopy(this.store)
yes in constructor we can also new Ctx(ctx, tx, gtxs, args..)
use this.state = ctx
this.tx = tx
like this

Comment (ratik21) [reply packages/runtime/src/ctx.ts:42]: > we will have these in types and we will have this class as Context. right ?

Exactly. Instead of having `runtime.ctx` as an obj, we can convert that to a class (which we can define in a new file, as done here). So this:
```
this.ctx = {
  state: cloneDeep(this.store), // state is a deep copy of store
  tx: tx,
  gtxs: gtxs,
  args: []
};
 ```   
 would update to:
 ```
 this.ctx = new Ctx(cloneDeep(this.store), tx, gtxs, [])
 ```
 And all ctx related functions would be in `Ctx` class.
 
 If you like it then we can update. Your call. 
 Otherwise we can just update the constructor here like you said above.

Comment (amityadav0) [reply packages/runtime/src/ctx.ts]: yes, i like it, updating to suggested one

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts]: just to confirm something, what about the errors thrown during execution other than `rejected_by_logic`? Let's say `txn Accounts` throws `index_out_of_bound` error (which happens during execution of `clear.teal`). We clear the state in these cases as well, Right?

Comment (amityadav0) [reply packages/runtime/src/runtime.ts]: No we don't.

Comment (ratik21) [reply packages/runtime/src/runtime.ts]: I am not sure, that's what i wanted to confirm here. Basically when we execute TEAL on blockchain (with index out of bound error, or something similar), does that constitute a tx `rejected_by_logic` there or not

Comment (amityadav0) [reply packages/runtime/src/runtime.ts]: https://developer.algorand.org/docs/features/asc1/teal/#getting-transaction-properties

Comment (amityadav0) [reply packages/runtime/src/runtime.ts]: updated as per network behaviour

File: packages/runtime/src/ctx.ts
Comment (ratik21) [inline packages/runtime/src/ctx.ts:135]:
```suggestion
   * @param index Index of current tx being processed in tx group
```

File: packages/runtime/src/ctx.ts
Comment (ratik21) [inline packages/runtime/src/ctx.ts:103]: Please update the params in description

File: packages/runtime/src/ctx.ts
Comment (ratik21) [inline packages/runtime/src/ctx.ts:231]: Please update the description

File: packages/runtime/src/ctx.ts
Comment (ratik21) [inline packages/runtime/src/ctx.ts:276]: Please also add the explanation (if not added somewhere). That we're doing  this because if any one tx in group fails, then it does not affect `runtime.store`, otherwise we just update `store` with `ctx` (if all transactions are executed successfully).

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:613]: just to confirm, i don't see the update  `this.ctx = {..}` to `this.ctx = new Ctx(..)` for app/asset creation & opt-in. Is that fine the previous way?

**Review by ratik21 — APPROVED** [review]
Left few comments. Please also check if we need to update `/docs/testing-teal.md` before merging.

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [reply packages/runtime/src/ctx.ts:103]: ok

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:613]: i think it is fine, otherwise test should fail.
-  asset creation, opt-in doesn't require ctx update.
- app creation, update and optin, we are updating ctx in functions

**Pull request merged**
